File: apps/desktop/src/renderer/services/runRedirect.ts
import { api } from "@/services/api";
import { resolveSidecarControlTargetForEngine } from "@/services/sidecarRouting";
import { useConversationStore } from "@/stores/conversation";
import type { InterveneAck } from "@agentcore/protocol-fold-kit";

export interface SubmitRunRedirectParams {
  executionId: string;
  runId: string;
  feedback: string;
}

/** 服务端回执：受理与否 + 一句话原因（`queued` 只是排队计数，不代表这次被受理）。 */
export type RunRedirectAck = InterveneAck & { queued: number };

/**
 * Redirect one in-flight worker (中间可见性 Phase 2a).
 *
 * Local turns route to the sidecar process (the cloud HTTP POST cannot reach the
 * in-process queue). The drive loop drains this queue on its next cancel poll:
 * the worker is cancelled outright, then re-run with the feedback — hot
 * `continue_run` from its salvaged transcript when continuable, else a
 * same-role `_redir` handoff from scratch (`runtime/delegate/drive_redirect.py`).
 * Nothing about it is「排队等下一步」——say so in any UI confirmation.
 *
 * 回执 `accepted` 说的是引擎有没有真收下：够不着这个 run 时它是 false，此时**什么都
 * 没发生**，界面不许说「已改方向」。
 */
export async function submitRunRedirect(
  conversationId: string,
  params: SubmitRunRedirectParams,
): Promise<RunRedirectAck> {
  const sidecarTarget = await resolveSidecarControlTargetForEngine(
    conversationId,
    useConversationStore.getState().byId[conversationId]?.executionVia,
  );
  if (sidecarTarget) {
    return window.sidecarApi.runRedirect({
      rootId: sidecarTarget.rootId,
      subpath: sidecarTarget.subpath,
      conversationId,
      executionId: params.executionId,
      runId: params.runId,
      feedback: params.feedback,
    });
  }
  return api.post<RunRedirectAck>(
    `/v1/conversations/${conversationId}/run-redirect`,
    {
      execution_id: params.executionId,
      run_id: params.runId,
      feedback: params.feedback,
    },
  );
}

/** Why a run's terminal outcome was recorded as accepted (跑一半改方向 Step 4).
 *  User-facing offer is only ``redirect_ignored``; the other two stay audit-only. */
export type RunOutcomeReason =
  | "deterministic_failure"
  | "redirect_ignored"
  | "recovery_ignored";

export interface AcceptRunOutcomeParams {
  /** The assistant message (turn) the run belongs to — scopes the audit trail. */
  messageId: string;
  runId: string;
  reason: RunOutcomeReason;
  executionId?: string;
  note?: string;
}

/**
 * Record the user's explicit accept of a run's terminal outcome (跑一半改方向 Step 4 · 忽略路径收口).
 *
 * Replaces the old frontend-only「忽略」(clearExecution) with a durable「用户主动接受此结果」row on
 * the SAME owner-scoped audit trail the run detail reads — so the acceptance survives reload and is
 * auditable. Cloud-only (like the audit read): the record lives with the turn's other audit rows.
 * Idempotent server-side; returns `recorded=false` if this run's outcome was already accepted.
 */
export async function acceptRunOutcome(
  conversationId: string,
  params: AcceptRunOutcomeParams,
): Promise<{ recorded: boolean }> {
  return api.post<{ ok: boolean; recorded: boolean; action: string }>(
    `/v1/conversations/${conversationId}/messages/${params.messageId}/accept-outcome`,
    {
      run_id: params.runId,
      reason: params.reason,
      execution_id: params.executionId,
      note: params.note,
    },
  );
}

File: apps/server/agentcore/api/routes/conversations/run_redirect.py
"""User-initiated worker redirect during an active delegate batch, plus the terminal-outcome
accept収口 (跑一半改方向 Step 4)."""

from typing import Any

from fastapi import APIRouter, Depends

from agentcore.api.dependencies import (
    AuthUser,
    get_agent_audit_repo,
    get_conversation_repo,
)
from agentcore.api.schemas import (
    AcceptRunOutcomeRequest,
    AcceptRunOutcomeResponse,
    SubmitRunRedirectRequest,
    SubmitRunRedirectResponse,
)
from agentcore.core.logging import get_logger
from agentcore.db.repositories import AgentAuditEventRepository, ConversationRepository
from agentcore.runtime.runs.intervene import accept_run_redirect

from ._helpers import _require_conversation_write

router = APIRouter(prefix="/conversations", tags=["conversations"])

logger = get_logger(__name__)


@router.post("/{conversation_id}/run-redirect", response_model=SubmitRunRedirectResponse)
async def submit_run_redirect(
    conversation_id: str,
    body: SubmitRunRedirectRequest,
    user: AuthUser,
    conv_repo: ConversationRepository = Depends(get_conversation_repo),
):
    """Queue a mid-flight redirect for one worker in the current delegate batch.

    The CEO is blocked inside ``delegate`` — this endpoint is the user直控 channel
    (WaveScheduler cancels the run, then hot ``continue_run`` or cold ``_redir``接手).

    The response says whether the engine actually took it (``accepted``): a run the
    live plan can't reach never enters the queue, and never reports success.
    """
    await _require_conversation_write(conversation_id, user.user_id, conv_repo._session)
    ack = accept_run_redirect(
        execution_id=body.execution_id,
        run_id=body.run_id,
        feedback=body.feedback,
        conversation_id=conversation_id,
    )
    if ack.accepted:
        logger.info(
            "run_redirect.queued",
            conversation_id=conversation_id,
            execution_id=body.execution_id,
            run_id=body.run_id,
            queued=ack.queued,
        )
    else:
        logger.info(
            "run_redirect.unreachable",
            conversation_id=conversation_id,
            execution_id=body.execution_id,
            run_id=body.run_id,
            reason=ack.reason,
        )
    return SubmitRunRedirectResponse(
        queued=ack.queued,
        accepted=ack.accepted,
        reason=ack.reason,
        detail=ack.detail,
    )


@router.post(
    "/{conversation_id}/messages/{message_id}/accept-outcome",
    response_model=AcceptRunOutcomeResponse,
)
async def accept_run_outcome(
    conversation_id: str,
    message_id: str,
    body: AcceptRunOutcomeRequest,
    user: AuthUser,
    conv_repo: ConversationRepository = Depends(get_conversation_repo),
    audit_repo: AgentAuditEventRepository = Depends(get_agent_audit_repo),
) -> AcceptRunOutcomeResponse:
    """Record the user's explicit accept of a run's terminal outcome (跑一半改方向 Step 4).

    Closes the loop for two audit-surfaced dead ends in a delegated turn: a
    ``deterministic_failure`` (retry徒劳) or a ``redirect_ignored`` (a「立即改此人」steer that could
    not apply mid-run). Replaces the old frontend-only ``clearExecution`` with a durable
    「用户主动接受此结果」row on the SAME append-only audit trail the run detail already reads — no
    new table, no new SSE event. Owner-scoped (对话归属校验防 IDOR) and idempotent per (turn, run):
    a second accept for the same run is a no-op (``recorded=false``).
    """
    await _require_conversation_write(conversation_id, user.user_id, conv_repo._session)
    rows = await audit_repo.list_for_turn(conversation_id=conversation_id, turn_id=message_id)
    # Idempotent per (turn, run): a repeated accept (double-click / retry) must not append twice.
    if any(r.action == "run.outcome_accepted" and r.run_id == body.run_id for r in rows):
        return AcceptRunOutcomeResponse(recorded=False)
    # Sort after the turn's existing audit events; inherit their trace_id for a joined timeline.
    seq = max((r.seq for r in rows), default=-1) + 1
    trace_id = next((r.trace_id for r in rows if r.trace_id), None)
    detail: dict[str, Any] = {"decision": "accepted", "reason": body.reason}
    if body.note:
        detail["note"] = body.note[:200]
    await audit_repo.append(
        user_id=user.user_id,
        conversation_id=conversation_id,
        turn_id=message_id,
        trace_id=trace_id,
        seq=seq,
        category="state",
        action="run.outcome_accepted",
        actor_kind="system",
        outcome="ok",
        execution_id=body.execution_id,
        run_id=body.run_id,
        detail=detail,
    )
    logger.info(
        "run_outcome.accepted",
        conversation_id=conversation_id,
        message_id=message_id,
        run_id=body.run_id,
        reason=body.reason,
    )
    return AcceptRunOutcomeResponse(recorded=True)
